mesos-client/README.md

```diff
@@ -0,0 +1,129 @@
+| Connection |
+ ------------
+      |
+      v
+ ------------
+| Connection | (2)
+| Handler    |     -> updates connection context
+ ------------
+      |
+      v
+ ------------
+| Data Bytes | (3)
+| Extractor  |
+ ------------
+      |
+      v
+ ------------
+| RecordIO   | (4)
+| Scanner    |
+ ------------
+      |
+      v
+ --------------
+| Event        | (5)
+| Deserializer |
+ --------------
+      |
+      v
+ --------------
+| Subscribed   | (6)  -> updates connection context
+| Handler      |
+ --------------
+      |
+      v
+ --------------
+| BroadcastHub | (6)
+ --------------
+  |  |  |  |  |
+  v  v  v  v  v
+```
+
+1. **Http Connection** mesos-v1-client uses akka-http low-level `Http.outgoingConnection()` to `POST` a [SUBSCRIBE](http://mesos.apache.org/documentation/latest/scheduler-http-api/#subscribe-1) request to mesos `api/v1/scheduler` endpoint providing framework info. The HTTP response is a stream in RecordIO format which is handled by the later stages
+2. **Connection Handler** handles connection HTTP response, saving `Mesos-Stream-Id`(see the description of the [SUBSCRIBE](http://mesos.apache.org/documentation/latest/scheduler-http-api/#subscribe-1) call) in client's _connection context_ object to later use it in mesos sink. Schedulers are expected to make HTTP requests to the leading master. If requests are made to a non-leading master a `HTTP 307 Temporary Redirect` will be received with the `Location` header pointing to the leading master. We update connection context with the new leader address and throw a `MesosRedicrectException` which is handled it in the `recover` stage by building a new flow that reconnects to the new leader
+3. **Data Byte Extractor** simply extracts byte data from the response body
+4. **RecordIO Scanner** Each stream message is encoded in RecordIO format, which essentially prepends to a single record (either JSON or serialized protobuf) its length in bytes: `[<length>\n<json string|protobuf bytes>]`. More about the format [here](http://mesos.apache.org/documentation/latest/scheduler-http-api/#recordio-response-format-1). RecordIO Scanner uses `RecordIOFraming.Scanner` from the [alpakka-library](https://github.com/akka/alpakka) to parse the extracted bytes into a complete message frame
+5. **Event Deserializer** Currently mesos-v1-client only supports protobuf encoded events/calls. Event deserializer uses [scalapb](https://scalapb.github.io/) library to parse the extracted RecordIO frame from the previous stage into a mesos [Event](https://github.com/apache/mesos/blob/master/include/mesos/scheduler/scheduler.proto#L36)
+6. **Subscribed Handler** parses the `SUBSCRIBED` event and updates the connection context with the returned `frameworkId`.
+7. **BroadcastHub** ad the end the flow is going through a broadcast hub. This allows for a dynamic "fan-out" streaming of events and avoids making multiple upstream connections to mesos when the source is materialized by multiple event consumers
```

Comment:
"at the end"

I would motion that we should leave the usage of BroadcastHub up to the user. It is problematic in that the subscribers may not see the first element, and, back-pressure is applied for the slowest of subscribers.

Reply:
I built it in because otherwise every materialisation of the `Source` would lead to a *new connection* to mesos being created. And we definitely don't want that. `BroadcastHub` is one way to avoid it - but I'm open to others!